content/blog/2023-12-11-A DSL for Metastability.md

```diff
@@ -0,0 +1,237 @@
+A general blueprint for these failures is as follows:
+1. The system is operating within a given environment.
+2. A trigger event happens. The trigger can come in different flavors: server crash, network partition, load surge, etc.
+3. The trigger pushes the system into a self-sustaining loop, in which the system is also "unavailable".
+4. Even after the triggering effect is remedied, the system remains in the self-sustaining loop.
+
+I have been trying to rigorously define metastability as a property of a suitably general system abstraction. We have a working definition that seems to capture the gist of it. However, since we are interested in actual computer systems that might experience metastability, it would be great if we could:
+1. express a generic "computer system", especially one that resembles the cloud in that it is made of clusters of homogenous machines,
+2. and analyze whether a given system - expressed in the previous step - "is metastable or not".
+
+The goal of this project was to design such a DSL, and to test its correctness. The desired outcome is a DSL with appropriate syntax and semantics, and an interpreter that receives a system expressed in this DSL and returns an execution trace. Such a trace can then be used to assess whether the system would suffer from metastability.
+
+# Design
+When I was first introduced to metastable failures, the main examples that were being used to study it were mostly about systems having a finite capacity to process stuff.
+Perhaps the most famous example is a retry storm, which roughly proceeds as follows:
+
+1. The system is operating under heavy load, such that it is about to be overloaded.
+2. A load surge happens, pushing the system into the overloaded state. This causes the system to drop some requests, and some requests timeout.
+3. The requests that are being timed out start retrying, and the retries keep growing as long as the overload state is sustained.
+4. After a certain threshold, even if the load surge is mitigated and the load returns to a level that *should* be safe according to the spec, the retries keep the system in the overload state.
+
+Based on this, my first attempt to capture this type of behavior was to imitate a method that researchers have used for other systems that demonstrate such *dynamic* behaviors: I modeled a computer system as a group of agents with finite capacity that interact according to predetermined rules.
+This finitude was captured through the following abstraction for an agent:
+
+1. An agent has a finite input queue and a finite output queue.
+2. An agent has finite receiving and sending rates on its input and output queues, respectively.
+3. An agent has a finite processing rate, where "processing" means taking something from the input queue and putting it into the output queue (note that there is no content to processing).
```

Comment:
By "no content to processing", do you mean that the only thing an agent can do with a message it receives is to put it, unchanged, into its output queue? It cannot, for example, produce N output messages for each input message?

EDIT: Never mind! I can see now that you have generalized this model in the paragraphs below. Probably just ignore this comment, but I'm leaving it here because I think it points out an interesting constraint on this notional baseline model.